src/v/iceberg/conversion/tests/iceberg_avro_tests.cc

```diff
@@ -1269,3 +1269,114 @@ INSTANTIATE_TEST_SUITE_P(
+    auto [schema, expected] = GetParam();
+
+    auto root = load_json_schema(schema);
+    ASSERT_THAT(root, AvroTypeIs(avro::AVRO_UNION));
+    ASSERT_THAT(expected, SizeIs(root->leaves()));
+
+    auto s_type = iceberg::type_to_iceberg(root);
+    ASSERT_TRUE(s_type.has_value());
+    ASSERT_THAT(s_type.value().fields, SizeIs(1));
+
+    ASSERT_TRUE(
+      std::holds_alternative<iceberg::struct_type>(
+        s_type.value().fields[0]->type));
+
+    const auto& u_type = std::get<iceberg::struct_type>(
+      s_type.value().fields[0]->type);
+
+    ASSERT_EQ(u_type.fields.size(), root->leaves());
+
+    for (auto i : std::views::iota(0UL, root->leaves())) {
+        auto label = iceberg::union_branch_label(
+          root->leafAt(i), u_type.fields[i]->type);
+        ASSERT_THAT(label, Eq(expected[i]));
+    }
+}
+
+using btc = branch_label_test_case;
+INSTANTIATE_TEST_SUITE_P(
+  UnionBranchLabelsTest,
+  BranchLabelTest,
+  Values(
+    btc{
+      .schema = R"J([ "long", "int" ])J",
+      .labels = {"long", "int"},
+    },
+    btc{
+      .schema = R"J([ "long", {"type": "map", "values": "int" } ])J",
+      .labels = {"long", "map"},
+    },
+    btc{
+      .schema = R"J([ "long", {"type": "array", "items": "int" } ])J",
+      .labels = {"long", "array"},
+    },
+    btc{
+      .schema
+      = R"J([ {"type": "record", "name": "Struct", "fields": [{"name": "a", "type": "int"}]}, "long" ])J",
+      .labels = {"Struct", "long"},
+    },
+    btc{
+      .schema
+      = R"J([ {"type": "enum", "name": "Enum", "symbols": ["FOO", "BAR"]}, "long" ])J",
+      .labels = {"Enum", "long"},
+    },
+    btc{
+      .schema
+      = R"J([ {"type": "fixed", "name": "Hash", "size": 16}, "long" ])J",
+      .labels = {"Hash", "long"},
+    },
+    btc{
+      .schema = R"J([ {"type": "string", "logicalType": "uuid"}, "long" ])J",
+      .labels = {"string", "long"},
+    },
+    btc{
+      .schema
+      = R"J([ {"type": "bytes", "logicalType": "decimal", "precision": 10, "scale": 2}, "long" ])J",
+      .labels = {"bytes", "long"},
+    },
+    btc{
+      .schema = R"J([ {"type": "int", "logicalType": "date"}, "long" ])J",
+      .labels = {"int", "long"},
+    },
+    btc{
+      .schema
+      = R"J([ {"type": "int", "logicalType": "time-millis"}, "long" ])J",
+      .labels = {"int", "long"},
+    },
+    btc{
+      .schema
+      = R"J([ {"type": "long", "logicalType": "time-micros"}, "string" ])J",
+      .labels = {"long", "string"},
+    },
+    btc{
+      .schema
+      = R"J([ {"type": "long", "logicalType": "timestamp-millis"}, "string" ])J",
+      .labels = {"long", "string"},
+    },
+    btc{
+      .schema
+      = R"J([ {"type": "long", "logicalType": "timestamp-micros"}, "string" ])J",
+      .labels = {"long", "string"},
+    },
+    btc{
+      .schema
+      = R"J([ {"type": "fixed", "name": "FixedDecimal", "size": 16, "logicalType": "decimal", "precision": 10, "scale": 2}, "long" ])J",
```

Comment:
why aren't we using the `name` here? Shouldn't we?

Reply:
Since these decision are hard to revert, can we make sure we don't miss any other cases?